client/client_test.go

```diff
@@ -33,6 +34,91 @@ func TestAddGetLeaf(t *testing.T) {
 	// TODO: Build a GetLeaf method and test a full get/set cycle.
 }
 
+func TestGetByIndex(t *testing.T) {
+	env, err := integration.NewLogEnv(context.Background(), 0, "TestGetByIndex")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer env.Close()
+	logID, err := env.CreateLog()
+	if err != nil {
+		t.Fatalf("Failed to create log: %v", err)
+	}
+
+	cli := trillian.NewTrillianLogClient(env.ClientConn)
+	client := New(logID, cli, testonly.Hasher, env.PublicKey)
+	// Add a few test leaves.
+	leafData := [][]byte{
+		[]byte("A"),
+		[]byte("B"),
+		[]byte("C"),
+	}
+
+	// TODO(gdbelvin): Replace with batch API.
+	// TODO(gdbelvin): Replace with AddSequenedLeaves API.
+	for _, l := range leafData {
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(timeout))
+		defer cancel()
+		if err, want := client.AddLeaf(ctx, l), codes.DeadlineExceeded; grpc.Code(err) != want {
```

Comment:
(Discussed offline too, comments for completeness.)

Re how to verify: we should look into returning some sort of promise. I'm personally against calling it SCT, as I don't want Trillian promises to be mistaken for "real" SCTs, but we definitely want something to play the corresponding part in Trillian.

Re quick integration / blocking: that doesn't solve the problem; outages and delays do happen, and when those happen the client will get an error and have nothing to show for it. I think dissociating Add from Verify (plus having a promise to show for adds) is key for good reliability in face of errors.